tests/core/test_simulation.py

```diff
@@ -0,0 +1,191 @@
+
+
+@pytest.fixture
+def mock_state_producer() -> Mock:
+    return create_autospec(StateProducer, instance=False)
+
+
+@pytest.fixture
+def patch_ticker() -> Iterable[Mock]:
+    with patch("tickit.core.management.schedulers.base.Ticker", autospec=True) as mock:
+        mock.return_value = AsyncMock(spec=Ticker)
+        mock.return_value.components = ["test"]
+        mock.return_value.time = SimTime(41)
+        mock.return_value.finished = asyncio.Event()
+        yield mock
+
+
+@pytest_asyncio.fixture
+async def patch_run_forever() -> AsyncGenerator:
+    with patch(
+        "tickit.core.management.schedulers.master.MasterScheduler.run_forever",
+        autospec=True,
+    ):
+        yield asyncio.create_task(asyncio.sleep(1))
+
+
+@pytest_asyncio.fixture
+async def mock_master_scheduler(
+    mock_wiring,
+    mock_state_consumer,
+    mock_state_producer,
+    patch_ticker,
+    patch_run_forever,
+) -> MasterScheduler:
+    mock_master_scheduler = MasterScheduler(
+        mock_wiring, mock_state_consumer, mock_state_producer
+    )
+    await mock_master_scheduler.setup()
+    return mock_master_scheduler
+
+
+@pytest.mark.asyncio
+async def test_running_only_scheduler(
+    mock_master_scheduler: MasterScheduler,
+) -> None:
+    tickit_simulation = TickitSimulation(
+        backend="internal", scheduler=mock_master_scheduler, components=None
+    )
+    assert tickit_simulation._components is None
+    assert tickit_simulation._scheduler is not None
+
+    await tickit_simulation.run()
+
+    tickit_simulation._scheduler.run_forever.assert_called_once()  # type: ignore
+
+
+@pytest_asyncio.fixture
+async def patch_component_run_forever() -> AsyncGenerator:
+    with patch(
+        "tickit.core.components.device_simulation.DeviceSimulation.run_forever",
+        autospec=True,
+    ):
+        yield asyncio.create_task(asyncio.sleep(1))
+
+
+@pytest.fixture
+def mock_components(patch_component_run_forever) -> Dict[ComponentID, Component]:
+    sink = Sink(name=ComponentID("test_sink"), inputs=dict())
+    source = Source(name=ComponentID("test_source"), inputs=dict(), value=42)
+    mock_components = {
+        ComponentID("test_sink"): sink(),
+        ComponentID("test_source"): source(),
+    }
+    return mock_components
+
+
+@pytest.mark.asyncio
+async def test_running_only_components(
+    mock_components: Dict[ComponentID, Component],
+) -> None:
+    tickit_simulation = TickitSimulation(
+        backend="internal", scheduler=None, components=mock_components
+    )
+
+    assert tickit_simulation._scheduler is None
+    assert tickit_simulation._components is not None
+
+    await tickit_simulation.run()
+
+    tickit_simulation._components
+
+    # device simulation run forever is called for each component in the dict
+    component = next(iter(tickit_simulation._components.values()))
+    assert component.run_forever.call_count == len(  # type: ignore
+        tickit_simulation._components
+    )
```

Comment:
It's not surprising, but it isn't nice. The patch is on a mutual call for components, the `device_simulation`, so it is called the same number of times as there are components. We're only checking after all the calls have been made so it is what I expect to see. I don't like it, but I failed to patch on the individual component level because of the data container approach with components, the actual thing I need to patch is created at runtime (it complained a lot).

Reply:
I've patched this with a mock I can call_count on.